In [1]:
import sys
!{sys.executable} -m pip install google-cloud-bigquery google-cloud-bigquery-storage pandas pyarrow db-dtypes

In [2]:
import pandas as pd
from google.cloud import bigquery
from google.oauth2 import service_account

credentials = service_account.Credentials.from_service_account_file("key.json") # lokalizacja pobranego klucza z punktu 1.4.

client = bigquery.Client(credentials=credentials, project=credentials.project_id)


In [3]:
query = """
SELECT *
FROM `bigquery-public-data.noaa_gsod.gsod2020`
    limit 10
"""

query_job = client.query(query)
query_result = query_job.result()
df = query_result.to_dataframe()

In [4]:
df

,stn,wban,date,year,mo,da,temp,count_temp,dewp,count_dewp,...,flag_min,prcp,flag_prcp,sndp,fog,rain_drizzle,snow_ice_pellets,hail,thunder,tornado_funnel_cloud
0,010030,99999,2020-10-12,2020,10,12,27.0,4,20.5,4,...,None,0.0,I,999.9,0,0,0,0,0,0
1,010350,99999,2020-11-08,2020,11,08,27.9,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0
2,010370,99999,2020-12-19,2020,12,19,22.3,4,16.6,4,...,None,0.0,I,999.9,0,0,0,0,0,0
3,010420,99999,2020-03-03,2020,03,03,21.5,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0
4,010420,99999,2020-02-18,2020,02,18,23.9,4,9999.9,0,...,*,0.0,I,999.9,0,0,0,0,0,0
5,010430,99999,2020-01-15,2020,01,15,29.2,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0
6,010430,99999,2020-02-01,2020,02,01,18.9,4,9999.9,0,...,*,0.0,I,999.9,0,0,0,0,0,0
7,010430,99999,2020-01-17,2020,01,17,25.1,4,9999.9,0,...,*,0.0,I,999.9,0,0,0,0,0,0
8,010430,99999,2020-03-18,2020,03,18,29.5,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0
9,010430,99999,2020-04-03,2020,04,03,24.0,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0


# Misja dodatkowa 3

### 3.1. Sprawdź, ile jest zapisanych wierszy z danymi.

In [5]:
query3_1 = """
select count(*) as count_rows from(
    SELECT  *
    FROM `bigquery-public-data.noaa_gsod.gsod2020`
    UNION ALL
    SELECT * FROM `bigquery-public-data.noaa_gsod.gsod2021`)
"""

df_3_1 = client.query(query3_1).to_dataframe()
df_3_1['count_rows'].loc[0]


np.int64(8145755)

### 3.2. Sprawdź, ile krajów jest uwzględnionych w danych.


In [6]:
query3_2 = """
        select count (Distinct country) as country_count from (SELECT *
        FROM `bigquery-public-data.noaa_gsod.gsod2020` AS gsod
                 INNER JOIN `bigquery-public-data.noaa_gsod.stations` AS st
                            ON st.usaf = gsod.stn and gsod.wban = st.wban
        UNION ALL
        SELECT *
        FROM `bigquery-public-data.noaa_gsod.gsod2021` AS gsod
                 INNER JOIN `bigquery-public-data.noaa_gsod.stations` AS st
                            ON st.usaf = gsod.stn and gsod.wban = st.wban )

        """

df_3_2= client.query(query3_2).to_dataframe()
df_3_2['country_count'].loc[0]


np.int64(237)

### 3.3. Sprawdź, w jaki sposób zapisywane są dzienne informacje dla poszczególnych lokalizacji.


In [10]:
query = """
SELECT *
FROM `bigquery-public-data.noaa_gsod.gsod2020`
    limit 10
"""

query_job = client.query(query)
query_result = query_job.result()
df = query_result.to_dataframe()
df

,stn,wban,date,year,mo,da,temp,count_temp,dewp,count_dewp,...,flag_min,prcp,flag_prcp,sndp,fog,rain_drizzle,snow_ice_pellets,hail,thunder,tornado_funnel_cloud
0,010030,99999,2020-10-12,2020,10,12,27.0,4,20.5,4,...,None,0.0,I,999.9,0,0,0,0,0,0
1,010350,99999,2020-11-08,2020,11,08,27.9,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0
2,010370,99999,2020-12-19,2020,12,19,22.3,4,16.6,4,...,None,0.0,I,999.9,0,0,0,0,0,0
3,010420,99999,2020-03-03,2020,03,03,21.5,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0
4,010420,99999,2020-02-18,2020,02,18,23.9,4,9999.9,0,...,*,0.0,I,999.9,0,0,0,0,0,0
5,010430,99999,2020-01-15,2020,01,15,29.2,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0
6,010430,99999,2020-02-01,2020,02,01,18.9,4,9999.9,0,...,*,0.0,I,999.9,0,0,0,0,0,0
7,010430,99999,2020-01-17,2020,01,17,25.1,4,9999.9,0,...,*,0.0,I,999.9,0,0,0,0,0,0
8,010430,99999,2020-03-18,2020,03,18,29.5,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0
9,010430,99999,2020-04-03,2020,04,03,24.0,4,9999.9,0,...,None,0.0,I,999.9,0,0,0,0,0,0


 0.   stn
 1.   wban
 2.   date - data obserwacji
 3.   year - rok obserwacji
 4.   mo   - miesiąc obserwacji
 5.   da   - dzień obserwacji
 6.   temp - średnia temperatura dobowa w stopniach Fahrenheita, zaokrąglona do części dziesiętnych *
 7.   count_temp  - liczba obserwacji
 8.   dewp - średni, dobowy punkt rosy w stopniach Fahrenheita, zaokrąglony do części dziesiętnych *
 9.   count_dewp  - liczba obserwacji
 10.  slp - średnie ciśnienie nad poziomem morza (mbar), zaokrąglone do części dziesiętnych *
 11.  count_slp - liczba obserwacji
 12.  stp - średnie ciśnienie na poziomie stacji pogodowej (mbar), zaokrąglone do części dziesiętnych *
 13.  count_stp - liczba obserwacji
 14.  visib - średnia widoczność w milach, zaokrąglona do części dziesiętnych **
 15.  count_visib   - liczba obserwacji
 16.  wdsp - średnia predkość wiatru w węzłach, zaokrąglona do części dziesiętnych **
 17.  count_wdsp   - liczba obserwacji
 18.  mxpsd - maksymalna, utrzymująca się prędkość wiatru w węzłach, zaokrąglona do części dziesiętnych **
 19.  gust - maksymalna prędkość podmuchu wiatru w węzłach, zaokrąglona do części dziesiętnych **
 20.  max - maksymalna temperatura zarejestrowana dla danego kraju i regionu w stopniach Fahrenheita, zaokrąglona do części dziesiętnych *
 21.  flag_max - wskaźnik informujący czy maksymalna temperatura została pobrana z raportu, czy wyliczona z danych godzinowych
 22.  min - minimalna temperatura zarejestrowana dla danego kraju i regionu w stopniach Fahrenheita, zaokrąglona do części dziesiętnych *
 23.  flag_min - wskaźnik informujący czy maksymalna temperatura została pobrana z raportu, czy wyliczona z danych godzinowych
 24.  prcp - całkowita suma opadów (również roztopiony śnieg) w calach i setnych częściach milimetra ***
 25.  flag_prcp - wskaźnik informujący ilości raportów zgłoszonych w ciągu dnia
        A - 1 raport 6-godzinnej ilości opadów
        B - suma 2 raportów 6-godzinnych
        C - suma 3 raportów 6-godzinnych
        D - suma 4 raportów 6-godzinnych
        E - 1 raport o 12-godzinnej ilości opadów
        F - suma 2 raportów 12-godzinnych
        G - 1 raport o 24-godzinnej ilości opadów
 26.  sndp - grbość pokrywy śnieżnej w calach, ostatni pomiar w ciągu dnia, jeżeli wykonano więcej niż 1 **
 27.  fog - wskaźnik informujący o występowaniu mgły w ciągu dnia
 28.  rain_drizzle - wskaźnik informujący o występowaniu mżawki w ciągu dnia
 29.  snow_ice_pellets - wskaźnik informujący o występowaniu krup śnieżnych bądź lodowych (mieszanka śniegu i lodu) w ciągu dnia
 30.  hail - wskaźnik informujący o występowaniu gradu w ciągu dnia
 31.  thunder - wskaźnik informujący o wystąpieniu burzy w ciągu dnia
 32.  tornado_funnel_cloud - wskaźnik informujący o wystąpieniu tornada bądź leja kondensacyjnego w ciągu dnia

\* brakujące wartości są zastąpione przez 9999.9
\** brakujące wartości są zastąpione przez 999.9
\*** brakujące wartości są zastąpione przez 99.99


### 3.4. Sprawdź, w jaki sposób zapisywane są wartości liczbowe.


wartości pomiarów zapisywane są w postaci liczb zmiennoprzecinkowych (float) zaokrąglone do pierwszego miejsca po przecinku, natomiast wartości opisyjące liczbe pomiarów są liczbami całkowitymi (integer)

### 3.5. Sprawdź, jaki przedział czasowy jest uwzględniony w danych. Dodatkowo porównaj zakresy czasowe dla różnych zmiennych pogodowych, takich jak temperatura i opady.


### 3.6. Sprawdź więcej informacji (co najmniej 5 różnych) o danych pogodowych. W tym celu nie wykonuj żadnych dodatkowych obliczeń. Wyniki przedstaw w formie krótkiego raportu.

# Część 4

### Przeanalizuj poniższe przypadki. Zastanów się, jakie dane potrzebujesz do każdego z nich, a następnie zapisz je w osobnych, jak najprostszych obiektach DataFrame. Na tym etapie nie wykonuj analiz ani obliczeń statystycznych. Wykonaj jedynie podstawowe uporządkowanie danych: wybierz potrzebne kolumny, zadbaj o poprawne typy danych (np. data), usuń duplikaty oraz jednoznacznie oznacz brakujące wartości. Gotowe dane z obiektów DataFrame zapisz w osobnych plikach CSV.

### 4.1. Chcemy posiadać podstawowe informacje o lokalizacjach pomiarów pogodowych (stacje) oraz krajach, tak aby dane były zrozumiałe dla człowieka i możliwe do dalszego przetwarzania.



### 4.2. Chcemy wygenerować podstawowe zestawienia dotyczące warunków pogodowych na świecie (np. temperatura, opady, wiatr) w ujęciu dziennym.



### 4.3. Chcemy poznać zjawiska ekstremalne w danych pogodowych poprzez uwypuklenie skrajnych wartości (np. bardzo wysokie/niskie temperatury, intensywne opady, silny wiatr).



### 4.4. Chcemy przygotować dane umożliwiające analizę zmian warunków pogodowych w czasie (np. dla wybranych lokalizacji lub zmiennych).



### 4.5. Zdefiniuj własny dodatkowy przypadek związany z danymi pogodowymi.



## Część 5

### Połącz ze sobą wszystkie dane otrzymane w części 4. Na ich podstawie utwórz jeden spójny zbiór danych i zapisz go jako oddzielny obiekt DataFrame. Jeżeli uznasz, że zasadne jest utworzenie więcej niż jednego zbioru, wykonaj to i uzasadnij swoją decyzję. Zadbaj o spójność struktury danych, w szczególności zgodność nazw kolumn, poprawność typów danych, brak duplikatów, jednoznaczne oznaczenie brakujących wartości. Wynik zapisz w pliku lub plikach CSV.

## Część 6

### Połącz ze sobą dane otrzymane w części 5 oraz dane znajdujące się w dodatkowym pliku CSV zawierającym informacje o produkcji rolniczej.
Wybierz jedno z poniższych źródeł danych:

### 6.1  Dane FAOSTAT (zalecane)

Dane dotyczące produkcji rolniczej dostępne w bazie FAOSTAT: https://www.fao.org/faostat/en/#data/QCL

Dane obejmują m.in.:

Area – nazwa kraju

Item – rodzaj uprawy

Year – rok

Element – typ wartości (np. Production, Yield)

Value – wartość

### 6.2. Crop Yield Prediction Dataset (Kaggle)

Dane dotyczące plonów rolniczych dostępne na platformie Kaggle: https://www.kaggle.com/datasets/patelris/crop-yield-prediction-dataset

Zbiór zawiera m.in.:

nazwę kraju,
rok,
rodzaj uprawy,
wielkość plonu lub produkcji.
Połącz dane pogodowe z danymi rolniczymi, zwracając szczególną uwagę na zgodność nazw krajów, dopasowanie zakresów czasowych, różnice w strukturze danych (np. dane dzienne vs roczne), brakujące wartości.

Nowy zbiór danych zapisz jako oddzielny obiekt DataFrame. Jeżeli uważasz, że należy stworzyć kilka takich obiektów, zrób to i zapisz swoje uzasadnienie.

Zadbaj o spójność struktury danych, w szczególności na zgodność nazw kolumn, poprawność typów danych, brak duplikatów, jednoznaczne oznaczenie brakujących wartości.

Wynik tego zadania zapisz w pliku lub plikach CSV.